core/course_controller/CourseController.py

```diff
@@ -2,12 +2,32 @@
 
 from core.local_data_classes import CourseFormData, CourseOverview, CourseRef, UserRef, CourseSectionRef, LabSectionRef
 from django.db import models
-from ta_scheduler.models import Course, CourseSection, LabSection
+from ta_scheduler.models import Course, CourseSection, LabSection, Semester
+
 
 
 class CourseController:
+
+    @staticmethod
+    def has_duplicate_course_id(course_code: str, semester: Semester, exclude_course_id: int | None = None) -> bool:
+        """
+        Checks if there is a duplicate course with the same course_code in the given semester.
+        Pre-conditions: course_code is a valid string, and semester is a valid Semester instance.
+        Post-conditions: Returns True if a duplicate exists, False otherwise.
+        Side-effects: None.
+        """
+        duplicate_course_query = Course.objects.filter(
+            course_code=course_code,
+            semester=semester,
+        )
+        # Exclude a specific course ID if provided (e.g., during updates)
+        if exclude_course_id:
+            duplicate_course_query = duplicate_course_query.exclude(id=exclude_course_id)
+        return duplicate_course_query.exists()
+
+
     @staticmethod
-    def save_course(course_data: CourseFormData, course_id: int | None = None) -> str:
+    def save_course(course_data: CourseFormData,semester_name: str,  course_id: int | None = None) -> str:
```

Comment:
course_code, not id

Reply:
this is where duplication checking would occur, if the course_code passed is None then we have to check if the code and semester in the course_data is identical to some other courses code/semester.